### ISSUE TITLE
YouTube Block Hidden after Move

### ISSUE THREAD

**Issue by AshboDev** [WordPress/gutenberg#6146]
## Issue Overview

When creating/editing a post, if you move a YouTube embed up or down, the video disappears from the editor, but is still there and will be saved when you publish. When the page is refreshed, or when you focus off the block it re-appears.

## Steps to Reproduce (for bugs)

1. Create post
2. Add some text, then a video block below it
3. Move the video block up and down a bit
4. It disappears

**Browser:** Chrome 66 Beta.
**WP Version:** 4.9.5 (New install, no plugins except Gutenberg)
**Gutenberg Version:** 2.6.0
**Server:** Laravel Valet (Nginx, MariaDB, PHP7.2)
**OS:** MacOS High Sierra 10.13.4
## Expected Behavior
The video moves up, preview intact for the user to visually see their page layout.

## Current Behavior
The preview disappears when the user moves the YouTube embed module, either by clicking the order arrows, or dragging and dropping.

## Possible Solution
Not too sure. I couldn't see any errors in the console.

## Screenshots / Video

https://gyazo.com/e62327e7c5b0832e513ea6dd7ebedde1

## Related Issues and/or PRs

Possibly #544 however that's about a re-render. This is a lack of rendering.

## Todos
- [ ] Tests
- [ ] Documentation

**Comment by jorgefilipecosta** [comment]
It is a really strange problem each time a block with a Sandbox goes "down"
( e.g: Move the block down, move the block after it up, move the block up and then undo ) the Sandbox becomes empty. The problem happens even if the HTML passed to the sandbox is not an iframe ( e.g: if we pass "<b> test content</b>" as HTML the problem also happens).
The iframe used inside Sandbox becomes empty when we move a block down when moving up everything stays ok.

**Comment by ianbelanger79** [comment]
I can confirm that this issue exists. I only see the embed previews disappear when moving the block down the page. Either by drag and drop or by using the sorting arrows. See gif below

![embed-bug](https://user-images.githubusercontent.com/34601694/45967453-103e9200-bffc-11e8-92e6-1c1eae26e77d.gif)

I have tested the Vimeo and YouTube embed blocks, both of which are affected by this bug.

**Comment by mtias** [comment]
This is an iframe re-render issue, if I recall correctly.

**Comment by earnjam** [comment]
In my testing this only occurs when moving the block down the page. Up does not cause the problem.

Also different browsers seem inconsistent on handling:
- Firefox will re-render automatically, so the empty space is temporary. 
- Chrome & Safari require a focus change and then refocusing on the block to re-render.

**Comment by youknowriad** [comment]
We explicitely remove iframes from the draggable object because if I remember properly, they'd be rerendered and sometimes freeze the draggable object.

I'm tempted to just close the issue.

**Comment by adamziel** [comment]
The problem stems from the following `WONTFIX`/`RESOLVED INVALID` WebKit and Firefox issues:
* https://bugs.webkit.org/show_bug.cgi?id=13574
* https://bugzilla.mozilla.org/show_bug.cgi?id=254144

Long story short:
* Imagine we have three blocks one under another: 1. Paragraph 2. Youtube Video 3. Columns
* Moving iframe DOM Node to a different place in DOM triggers a reload.
* Moving the Youtube Video up leaves the DOM Node untouched. Instead, React moves the Paragraph block DOM Node below the Youtube video block.
* Moving the Youtube Video down leaves the Columns block DOM Node untouched, and moves the Youtube Video DOM Node below it.

The only possible fix here (assuming we're retaining the iframe sandbox) would be detecting which of the two repositioned blocks contains an iframe, and retaining it's DOM Node in place. But then both nodes could be videos. And there's also a drag handle to move the block around with a mouse. And what about cut/paste scenario?

I don't think this one is truly fixable, so I'm going to close this issue. Feel free to reopen if you see a way to address it.

**Comment by jorgefilipecosta** [comment]
> Moving an <iframe /> DOM Node to a different place in DOM triggers a reload.

Hi @adamziel, if the iframe reloads how does it become empty on the second reload instead of reloading displaying correctly?

**Comment by adamziel** [comment]
@jorgefilipecosta Ah good point! This is in fact super interesting:

* Reloading itself results in an empty iframe - we do all the work to initialize the contents in `Sandbox.trySandbox` which is called `onLoad`
* Moving the block manually using chrome devtools triggers `onLoad` :+1:
* Moving the block using BlockMover does not trigger `onLoad` :-1: - only after the iframe is clicked or somehow interacted with the `onLoad` is getting triggered (root cause)
* `<Sandbox />` component is unaware that the parent block was moved

So - there is absolutely no way around reloading the iframe (and restarting the video being played), but we should be able to do something with the empty iframe issue. I am investigating right now. I am not sure if it's possible to easily address the root cause since it may be related to the way React updates the DOM, but maybe we can work around it using some mutation observers?

**Comment by adamziel** [comment]
We're hitting the following React issue: https://github.com/facebook/react/issues/18752 
Fortunately there is an easy way around (registered onLoad event handler using `addEventListener`)

**Comment by adamziel** [comment]
This PR should fix it: https://github.com/WordPress/gutenberg/pull/21916
